Fix has one association multiple savings

briu opened the pull request:
### Motivation / Background
"Fixes #[48077](https://github.com/rails/rails/issues/48077)"

### Detail

When object is saved with `belongs_to` association, that associated to object with `has_one` relation, object actually saved   twice. For example:
```
class Post < ActiveRecord::Base
  belongs_to :user
  belongs_to :poll

  after_commit :test_after_commit
  # after_save :test_after_commit saved_changes here are exists
  after_save :test_after_save

  attr_reader :after_save_counter

  def test_after_commit
    # saved_changes here are blank
    if saved_change_to_release_stage?(to: 'published')
      user.increment(:published_posts_count)
    end
  end

  def test_after_save
    # invokes twice
    @after_save_counter ||= 0
    @after_save_counter += 1
  end
end

class Poll < ActiveRecord::Base
  has_one :post
end

post = Post.new(
  user: user,
  title: 'yoyo',
  release_stage: 'published'
)

post.poll = Poll.new(multiple: true)

post.save
```

First poll object is saved, then AR try to save post from `save_has_one_association`, after that invokes `save` method second time and this second `save` call clears saved_changes.

I decided to add variable that show that object already saving and prevent second save from `save_has_one_association` callback

jorgemanrubia committed:
Freeze casted values instead of original values in database queries

This deals with a problem introduced in #7743ab95b8e15581f432206245c691434a3993d1a751b9d451170956d59457a9R8
that was preventing query `Class` serialized attributes. Duplicating the original
`Class` argument generates an anonymous class that can't be serialized as YAML.

This change makes query attributes hasheable based on their frozen casted values
to prevent the problem.

This solution is based on an idea by @matthewd from https://github.com/rails/rails/issues/47338#issuecomment-1424402777.

alpaca-tc committed:
Deprecate `deferrable: true` option of `add_foreign_key`.

`deferrable: true` is deprecated in favor of `deferrable: :immediate`, and
will be removed in Rails 7.2.

Because `deferrable: true` and `deferrable: :deferred` are hard to understand.
Both true and :deferred are truthy values.
This behavior is the same as the deferrable option of the add_unique_key method, added in #46192.

*Hiroyuki Ishii*

alpaca-tc committed:
The deferrable foreign key can be passed to `t.references`.

Fixed a bug in which deferrable foreign_key was ignored when passed to `t.references`.

g-pavlik committed:
Fix Rack::Test::UploadedFile.new to work with StringIO

Instantiating Rack::Test::UplaodedFile with StringIO
no longer causes an exception.

nirebu committed:
Do not allocate the first character when checking for relative paths

This is the same optimization applied in
https://github.com/rails/rails/commit/a63ae913dfbb3813256144e52ae1063ac59edba4
which I proposed in https://github.com/rails/rails/pull/47714

Here's the benchmark:

require "bundler/inline"

ROOT_STRING = '/'
TEST_PATH = "/some/path"

gemfile(true) do
  source "https://rubygems.org"

  gem "benchmark-ips"
end

Benchmark.ips do |x|
  x.report("path[0]") do
   TEST_PATH[0] != ROOT_STRING
  end

  x.report("path.start_with?") do
   TEST_PATH.start_with?(ROOT_STRING)
  end

  x.compare!
end

Warming up --------------------------------------
             path[0]   942.044k i/100ms
    path.start_with?     1.556M i/100ms
Calculating -------------------------------------
             path[0]      9.463M (± 0.9%) i/s -     48.044M in   5.077358s
    path.start_with?     15.611M (± 0.2%) i/s -     79.352M in   5.083056s

Comparison:
    path.start_with?: 15611192.8 i/s
             path[0]:  9463245.0 i/s - 1.65x  slower

john-h-k committed:
Add `intersects?` to Relation

Ruby 3.1 added `intersects?` which is equivalent to `(a & b).any?`. Rubocop added a corresponding cop, `Style/ArrayIntersect`, which transforms the old style to use `intersect?`. Unfortunately as `intersects?` is not delegated on `CollectionProxy`, this leads to false positives that need to be disabled for no good reason other than the fact the method isn't delegated.

This PR add delegation of `intersects?` to `Relation` which fixes this.

joshuay03 committed:
[Fix #45389] updated_at not updating in before_update callback

This commit ensures that the updated_at timestamp is re-written
once all update callbacks have been run so that mutations made within
those callbacks are taken into account on persistence.

ghiculescu committed:
Fix error in `ActiveRecord::Migration.check_pending!`

Since https://github.com/rails/rails/commit/e9d835f05a90d6da03f69e1ecb59bc3f38edb979 this method has not worked, it has failed with this error:

```
NoMethodError: undefined method `pending_migrations' for #<ActiveRecord::MigrationContext:0x00005598468f8e58>
Did you mean?  pending_migration_versions
```

I agree with Eileen that eventually we could deprecate it, but to get it working I just reversed the changes in https://github.com/rails/rails/commit/e9d835f05a90d6da03f69e1ecb59bc3f38edb979 and added a regression test.

p8 committed:
Unlink headings for alternative libs in assets guide [ci-skip]

By linking to the libraries from the headings the anchor links in the
sidebar link directly to the libraries instead of the library's section.
This behaviour is unexpected.
We can link to the libraries in the section instead.

fatkodima committed:
Allow `insert_all`/`upsert_all` to use indexes with columns not in the same order as in `:unique_by`

agrobbin committed:
add a section on exclusion constraints to the AR PostgreSQL guide

joshuay03 committed:
Consistently raise an `ArgumentError` if the `ActiveSupport::Cache` key is blank

briu committed:
prevent double savings when saved through has_one association

shouichi committed:
document AD::FlashHash#delete [skip ci]

Co-authored-by: Seonggi Yang <seonggi.yang@gmail.com>
Co-authored-by: Ryohei UEDA <ueda@anipos.co.jp>
Co-authored-by: oljfte <oljfte@gmail.com>

briu committed:
add spec

byroot committed:
Make Adapter#exec_query clear the query cache

It's public API and we can't assume whether the query is read only
so we should clear the cache.

To perform read only queries, `select` and `select_all` can be used.

byroot committed:
Rename `uses_transactions` to `materialize_transactions`

It's hard to understand what `use_transactions` does.

shivamsinghchahar committed:
Remove noise from actionmailer callbacks test

shivamsinghchahar committed:
Add tests for AJ to verify queue name passed as symbol. Ref: #48034

briu committed:
init variable inside method

carlosantoniodasilva committed:
Use the skip forgery protection helper in Active Storage guide

The helper method is essentially a wrapper around skipping the callback,
and it simplifies the doc example.

[ci skip]

nvasilevski commented on activerecord/lib/active_record/autosave_association.rb:
```diff
@@ -272,6 +273,10 @@ def changed_for_autosave?
     end
 
+    def already_trying_to_save?
+      @saving ||= false
```
It's not going to lead to a problem but I find memoization here being unnecessary as it will still assign `false` to @savings` if it's nil or has already been assigned to `false`

I think we can just rely on the variable value being truthy/falsey
```suggestion
      @saving
```

briu replied:
yep, I've done it in this way because of CI warnings

moved instantiation to `init_internals` method

nvasilevski commented on activerecord/lib/active_record/autosave_association.rb:
```diff
@@ -366,6 +371,12 @@ def normalize_reflection_attribute(indexed_attribute, reflection, index, attribu
       end
 
+      def around_save_mark_as_saving_record
+        @saving = true
```
Perhaps it would be a good practise to name this variable as verbose as possible and perhaps prefix it with an underscore so there is less risk of a collision with instance variables defined by the application. Perhaps something like `@_marked_as_saving`

briu replied:
agreed, thanks for suggestion

shivamsinghchahar committed:
Fix flaky test in AR - token for

pass time arg to @user.touch

aharpole committed:
In older migrations, allow column collation to be explicitly specified

wpank84 committed:
Update migration example to showcase database-agnostic raw SQL execution

jonathanhefner committed:
Support :message_pack as a cache serializer format

This commit adds support for `:message_pack` as an option for
`config.active_support.cache_format_version`.

Cache entries written using the `6.1` or `7.0` formats can be read when
using the `:message_pack` format. Additionally, cache entries written
using the `:message_pack` format can now be read when using the `6.1` or
`7.0` format. These behaviors makes it easy to migrate between formats
without invalidating the entire cache.

joshuay03 committed:
[Fixes #48080] Fix broken distinct in `ActiveRecord::Calculations#ids` with `includes(...).order(...)`

byroot committed:
`Object#deep_dup` no longer duplicate named classes and modules.

It makes very little sense to duplicate named classes and modules.

Before:

```ruby
hash = { class: Object, module: Kernel }
hash.deep_dup # => {:class=>#<Class:0x00000001063ffc80>, :module=>#<Module:0x00000001063ffa00>}
```

After:

```ruby
hash = { class: Object, module: Kernel }
hash.deep_dup # => {:class=>Object, :module=>Kernel}
```

byroot committed:
Revert "Merge pull request #47352 from basecamp/ar-freeze-cached"

This reverts commit 2c20f90ebae3fdd4d7e3351aeaffc4ad2c472069, reversing
changes made to 912096d4ce930b8e7e5d91e0c86bae2091fda0e4.

byroot committed:
Add a test for querying serialized attributes containing Module

Extracted from https://github.com/rails/rails/pull/47352

Mangara committed:
Improve queue adapter name extraction for logs

We recently made improvements to the name extraction for `MyJob.queue_adapter_name` (see https://github.com/rails/rails/pull/47995 and https://github.com/rails/rails/pull/48003). I had overlooked that we also extract the queue adapter name in a similar (but not identical) way for log messages, so this change aligns the two.

adrianna-chang-shopify committed:
Rename MySQL::DatabaseStatements => Mysql2::DatabaseStatements

adrianna-chang-shopify committed:
Remove `AbstractMysqlAdapterTest`

This test was added in 8f5095a to ensure that there was coverage
for `AbstractMysqlAdapter#execute`, but 63c0d6b refactored `#execute`
to be defined at the Abstract adapter level, and rely on concrete MySQL
adapters to implement `#raw_execute`.

This test won't pass without having the `ExampleMysqlAdapter` implement
`#raw_execute`, but this test is obsolete now that AbstractMysql2Adapter
doesn't implement `#execute` directly.

adrianna-chang-shopify committed:
Move shared database logic to MySQL::DatabaseStatements

adrianna-chang-shopify committed:
Use String instead of Regexp when checking `#include?` for Trilogy conn errors

byroot committed:
Update pg gem to 1.5.3

Ref: https://github.com/ged/ruby-pg/issues/530

Fix segfault witnessed in https://buildkite.com/rails/rails/builds/96048#0187c9f9-fafa-4795-a52f-163071126d61

a5-stable committed:
allow rewhere to pass nil and return unscope(:where)

shivamsinghchahar committed:
Fix typo in activejob tests

changes `know` -> `known` in activejob serializers test

briu committed:
rename variable to @_marked_as_saving

briu committed:
instantiate variable in init_internals method

briu committed:
Merge branch 'main' into fix_has_one_association_multiple_savings

briu committed:
return variable without instantiation in already_trying_to_save? method

briu commented:
@nvasilevski seems that I rebased main in a wrong way and currently I have tons of conflicts when I tried to squash commits into one. 

Is it ok to close this PR and just reopen it with current code changes?

briu commented:
reopen this PR [here](https://github.com/rails/rails/pull/48130) because of wrong rebase and a lot of conflicts with main in current PR
